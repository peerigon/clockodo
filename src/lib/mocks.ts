// These date helpers are not meant to replace an actual date library.
// They are inaccurate and sometimes just wrong, see https://gist.github.com/timvisee/fcda9bbdff88d45cc9061606b4b923ca
// They should only be used for generating mock data.

import { faker } from "@faker-js/faker";

/**
 * Sets the seed for Faker.js (which is used by the mocks).
 *
 * @see https://github.com/Marak/Faker.js
 */
export const setFakerSeed = (seed: number) => {
  faker.seed(seed);
};

/** The number of milliseconds on a typical day. */
export const ONE_DAY = 24 * 60 * 60 * 1000;

/** The number of milliseconds on a typical year. */
export const ONE_YEAR = 356 * ONE_DAY;

const MAX_ITERATION_COUNT = 10_000;

export const startOfDay = (date: Date) => {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate());
};

export const endOfMonth = (date: Date) => {
  return new Date(date.getFullYear(), date.getMonth() + 1, 0, 23, 59, 59, 999);
};

export const endOfYear = (dateTime: Date) => {
  return new Date(dateTime.getFullYear() + 1, 0, 0, 23, 59, 59, 999);
};

export const startOfNextDay = (date: Date) => {
  const nextDay = new Date(startOfDay(date).getTime() + 1.5 * ONE_DAY);

  return new Date(nextDay.getFullYear(), nextDay.getMonth(), nextDay.getDate());
};

const generateWithMaxDuplicates = <Value>({
  count,
  maxDuplicates = 1,
  generate,
}: {
  count: number;
  maxDuplicates?: number;
  generate: () => Value;
}) => {
  const accepted = new Map<Value, Array<Value>>();
  let acceptedCount = 0;

  for (let i = 0; acceptedCount < count; i++) {
    if (i === MAX_ITERATION_COUNT) {
      throw new Error(
        `Couldn't generate enough unique values before reaching the max iteration count.
This usually happens when the generated values create too many conflicts (e.g. too many dates on a small date range).`,
      );
    }
    const value = generate();
    const duplicates = accepted.get(value) ?? [];

    if (duplicates.length >= maxDuplicates) continue;

    duplicates.push(value);
    accepted.set(value, duplicates);
    acceptedCount++;
  }

  return [...accepted.values()].flat();
};

export const generateRandomDateTimes = ({
  count,
  between: [from, to],
  maxDuplicates = 1,
}: {
  count: number;
  between: readonly [Date, Date];
  maxDuplicates?: number;
}) => {
  return generateWithMaxDuplicates({
    count,
    maxDuplicates,
    generate: () => {
      return faker.date.between({ from, to }).getTime();
    },
  }).sort();
};

export const generateRandomDates = ({
  count,
  between: [from, to],
  maxDuplicates = 1,
}: {
  count: number;
  between: readonly [Date, Date];
  maxDuplicates?: number;
}) => {
  return generateWithMaxDuplicates({
    count,
    maxDuplicates,
    generate: () => {
      const randomDate = faker.date.between({ from, to });

      return new Date(
        randomDate.getFullYear(),
        randomDate.getMonth(),
        randomDate.getDate(),
      ).getTime();
    },
  }).sort();
};

export const generateRandomMonths = ({
  count,
  between: [from, to],
  maxDuplicates = 1,
}: {
  count: number;
  between: readonly [Date, Date];
  maxDuplicates?: number;
}) => {
  return generateWithMaxDuplicates({
    count,
    maxDuplicates,
    generate: () => {
      const randomDate = faker.date.between({ from, to });

      return new Date(
        randomDate.getFullYear(),
        randomDate.getMonth(),
      ).getTime();
    },
  }).sort();
};

export const toPairs = <Item>(array: Array<Item>) => {
  if (array.length % 2 !== 0) {
    throw new Error("Cannot create pairs: Array length must be divisible by 2");
  }

  const count = array.length / 2;
  const pairs: Array<[Item, Item]> = Array.from({ length: count });

  for (let i = 0; i < count; i++) {
    pairs[i] = [array[i * 2], array[i * 2 + 1]];
  }

  return pairs;
};
