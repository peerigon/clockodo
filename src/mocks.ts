import { faker } from "@faker-js/faker";

export * from "./models/absence.mocks.js";
export * from "./models/customer.mocks.js";
export * from "./models/entry.mocks.js";
// export * from "./models/entryGroup.mocks.js";
export * from "./models/holidayscarry.mocks.js";
export * from "./models/holidaysquota.mocks.js";
export * from "./models/lumpsumService.mocks.js";
export * from "./models/nonbusinessDay.mocks.js";
export * from "./models/project.mocks.js";
// export * from "./models/service.mocks.js";
export * from "./models/targethours.mocks.js";
export * from "./models/team.mocks.js";
export * from "./models/user.mocks.js";
// export * from "./models/userReport.mocks.js";
export * from "./models/worktimeRegulation.mocks.js";
export * from "./models/workTimes.mocks.js";

/**
 * Sets the seed for Faker.js (which is used by the mocks).
 *
 * @see https://github.com/Marak/Faker.js
 */
export const setFakerSeed = (seed: number) => {
  faker.seed(seed);
};
