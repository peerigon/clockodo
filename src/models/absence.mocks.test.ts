import { describe, expect, test } from "vitest";
import { createAbsenceMocks } from "./absence.mocks.js";

describe("createAbsenceMocks()", () => {
  test("It creates realistic mocks", () => {
    expect(createAbsenceMocks({ count: 20 })).toMatchSnapshot();
  });

  test("Creates absences on a single day", () => {
    for (let i = 0; i < 20; i++) {
      const [absence] = createAbsenceMocks({
        count: 1,
        dateSinceBetween: [new Date(2020, 0, 1), new Date(2020, 0, 1)],
      });

      expect(absence).toMatchObject({
        dateSince: "2020-01-01",
        dateUntil: "2020-01-01",
      });
    }
  });
});
