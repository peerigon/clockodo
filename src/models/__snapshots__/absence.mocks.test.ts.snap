// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAbsenceMocks() It creates realistic mocks 1`] = `
Array [
  Object {
    "approvedBy": 1,
    "countDays": 5,
    "countHours": null,
    "dateApproved": "2019-09-25",
    "dateEnquired": "2019-09-24",
    "dateSince": "2020-01-26",
    "dateUntil": "2020-01-31",
    "id": 0,
    "note": "nam commodi ad animi corporis",
    "sickNote": null,
    "status": 1,
    "type": 1,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": "2019-12-02",
    "dateSince": "2020-03-01",
    "dateUntil": "2020-03-01",
    "id": 1,
    "note": "eum dolores",
    "sickNote": null,
    "status": 3,
    "type": 2,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 9,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-03-12",
    "dateUntil": "2020-03-24",
    "id": 2,
    "note": null,
    "sickNote": false,
    "status": 3,
    "type": 4,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 1,
    "countHours": null,
    "dateApproved": "2019-12-28",
    "dateEnquired": "2019-12-26",
    "dateSince": "2020-04-07",
    "dateUntil": "2020-04-10",
    "id": 3,
    "note": "fuga pariatur voluptates non beatae",
    "sickNote": false,
    "status": 1,
    "type": 5,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 1,
    "countHours": null,
    "dateApproved": "2019-11-21",
    "dateEnquired": null,
    "dateSince": "2020-05-03",
    "dateUntil": "2020-05-05",
    "id": 4,
    "note": "unde ut corrupti eum nostrum consequatur aliquam nostrum commodi",
    "sickNote": null,
    "status": 1,
    "type": 6,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 5.04,
    "dateApproved": "2020-04-01",
    "dateEnquired": "2020-03-28",
    "dateSince": "2020-05-24",
    "dateUntil": "2020-05-25",
    "id": 5,
    "note": null,
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": "2020-03-24",
    "dateSince": "2020-05-29",
    "dateUntil": "2020-05-29",
    "id": 6,
    "note": "debitis velit eos sequi dignissimos sint dolorem quasi",
    "sickNote": null,
    "status": 0,
    "type": 8,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 4.8,
    "dateApproved": "2019-11-19",
    "dateEnquired": "2019-11-17",
    "dateSince": "2020-06-04",
    "dateUntil": "2020-06-05",
    "id": 7,
    "note": null,
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 1,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-06-05",
    "dateUntil": "2020-06-09",
    "id": 8,
    "note": "ad laudantium",
    "sickNote": null,
    "status": 3,
    "type": 10,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 7,
    "countHours": null,
    "dateApproved": "2019-12-30",
    "dateEnquired": null,
    "dateSince": "2020-06-15",
    "dateUntil": "2020-06-25",
    "id": 9,
    "note": "corporis nisi doloremque consectetur aliquam suscipit aliquid corrupti",
    "sickNote": null,
    "status": 1,
    "type": 11,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 1,
    "countHours": null,
    "dateApproved": "2020-06-28",
    "dateEnquired": null,
    "dateSince": "2020-07-09",
    "dateUntil": "2020-07-10",
    "id": 10,
    "note": "vero sed dolore aliquid optio adipisci nemo nostrum",
    "sickNote": null,
    "status": 1,
    "type": 12,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": null,
    "countHours": 4.52,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-07-13",
    "dateUntil": "2020-07-20",
    "id": 11,
    "note": "id sequi illum dolore quos facere",
    "status": 4,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": "2020-06-27",
    "dateSince": "2020-08-04",
    "dateUntil": "2020-08-04",
    "id": 12,
    "note": "assumenda ipsa quis iusto quos",
    "sickNote": null,
    "status": 4,
    "type": 14,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 1,
    "countHours": null,
    "dateApproved": "2020-02-04",
    "dateEnquired": "2020-01-31",
    "dateSince": "2020-08-17",
    "dateUntil": "2020-08-17",
    "id": 13,
    "note": "saepe exercitationem neque atque dolore in totam",
    "sickNote": null,
    "status": 1,
    "type": 15,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 7.87,
    "dateApproved": "2020-05-25",
    "dateEnquired": null,
    "dateSince": "2020-09-07",
    "dateUntil": "2020-09-09",
    "id": 14,
    "note": "ullam nisi culpa tempore at consequuntur magni unde doloribus",
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": null,
    "countHours": 3.66,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-09-11",
    "dateUntil": "2020-09-11",
    "id": 15,
    "note": null,
    "status": 0,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": null,
    "countHours": 0.48,
    "dateApproved": null,
    "dateEnquired": "2020-07-16",
    "dateSince": "2020-09-17",
    "dateUntil": "2020-09-17",
    "id": 16,
    "note": "recusandae velit excepturi ipsa esse ab eligendi molestiae soluta",
    "status": 0,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": "2020-07-07",
    "dateEnquired": "2020-07-03",
    "dateSince": "2020-09-20",
    "dateUntil": "2020-09-20",
    "id": 17,
    "note": null,
    "sickNote": true,
    "status": 1,
    "type": 5,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 3.29,
    "dateApproved": "2020-05-16",
    "dateEnquired": "2020-05-13",
    "dateSince": "2020-09-25",
    "dateUntil": "2020-09-27",
    "id": 18,
    "note": "vitae ratione est odio officiis",
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 22,
    "countHours": null,
    "dateApproved": "2020-06-24",
    "dateEnquired": null,
    "dateSince": "2020-10-15",
    "dateUntil": "2020-11-06",
    "id": 19,
    "note": "veritatis accusamus cupiditate cupiditate",
    "sickNote": null,
    "status": 1,
    "type": 7,
    "usersId": 0,
  },
]
`;
